src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -902,70 +901,26 @@ where
-                    .await;
-
-                let mut handle = match handle {
-                    Ok(stuff) => stuff,
-                    Err(e) => {
-                        panic!("Invalid usage of the persist client for source {source_id} status history shard: {e:?}");
-                    }
-                };
-
-                let mut recent_upper = handle.upper().clone();
-                let mut append_ts = timestamp;
-                'retry_loop: loop {
-                    // We don't actually care so much about the timestamp we append at; it's best-effort.
-                    // Ensure that the append timestamp is not less than the current upper, and the new upper
-                    // is past that timestamp. (Unless we've advanced to the empty antichain, but in
-                    // that case we're already in trouble.)
-                    for t in recent_upper.elements() {
-                        append_ts.join_assign(t);
-                    }
-                    let mut new_upper = Antichain::from_elem(append_ts.step_forward());
-                    new_upper.join_assign(&recent_upper);
-
-                    let updates = vec![
-                        ((SourceData(Ok(row.clone())), ()), append_ts, 1i64)
-                    ];
-                    let cas_result = handle.compare_and_append(updates, recent_upper.clone(), new_upper).await;
-                    match cas_result {
-                        Ok(Ok(Ok(()))) => break 'retry_loop,
-                        Ok(Ok(Err(Upper(upper)))) => {
-                            recent_upper = upper;
-                        }
-                        Ok(Err(e)) => {
-                            panic!("Invalid usage of the persist client for source {source_id} status history shard: {e:?}");
-                        }
-                        Err(e) => {
-                            trace!("compare_and_append in update_status failed: {e}");
-                        }
-                    }
-                }
+            let status_shard = if let Some(status_shard) = storage_metadata.status_shard {
+                status_shard
+            } else {
+                // If there's no status shard, this operator has no work to do.
+                return
+            };
 
-                handle.expire().await
+            let persist_client = {
+                let mut persist_clients = persist_clients.lock().await;
+                persist_clients
+                    .open(storage_metadata.persist_location.clone())
+                    .await
+                    .expect("error creating persist client for Healthchecker")
             };
 
             if is_active_worker {
                 info!("Health for source {source_id} initialized to: {last_reported_status:?}");
                 let now_ms = now();
                 let row = healthcheck::pack_status_row(source_id, last_reported_status.name(), last_reported_status.error(), now_ms);
```

Comment:
I could also pull this out.  What do you think?  I'm not sure how we expect the metadata field to evolve

Reply:
Mildly in favour, since it seems like the two impls really ought to have the same behaviour, but I don't feel strongly.